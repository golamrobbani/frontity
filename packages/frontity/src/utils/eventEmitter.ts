import { EventEmitter } from "events";
import TypedEmitter from "typed-emitter";

interface MessageEvents {
  "cli:create:error": (error: Error) => void;
  "cli:create:subscribe": (msg: string, action: Promise<any>) => void;
  "cli:create:message": (msg: string, step: Promise<any>) => void;

  "cli:subscribe": (msg: string, action: Promise<any>) => void;
  "cli:subscribe:error": (error: Error) => void;
}

const emitter = new EventEmitter() as TypedEmitter<MessageEvents>;

export { emitter };
