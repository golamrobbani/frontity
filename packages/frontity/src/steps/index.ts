import fetch from "node-fetch";
import { emitter } from "../utils/eventEmitter";

const isEmailValid = (email: string): boolean =>
  /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,63}$/i.test(email);

export default async (email: string) => {
  const emit = (message: string, step?: Promise<any>) => {
    if (emitter) emitter.emit("cli:create:subscribe", message, step);
  };

  let step: Promise<any>;

  try {
    if (!isEmailValid(email))
      throw new Error("Email not valid. Please enter a valid email.");

    step = fetch(
      "https://hook.integromat.com/gm0b502jo5acuhzko7gszx0kd9r52ofi",
      {
        method: "POST",
        body: JSON.stringify({
          event: "frontity-subscribe",
          email: email.toLowerCase()
        }),
        headers: { "Content-Type": "application/json" }
      }
    );
    emit("Subscribing to Frontity", step);
    await step;
  } catch (error) {
    if (emitter) emitter.emit("cli:create:error", error);
    else throw error;
  }
};
