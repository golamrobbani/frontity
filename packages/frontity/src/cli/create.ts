import { resolve } from "path";
import ora from "ora";
import chalk from "chalk";
import { prompt, Question } from "inquirer";
import create from "../commands/create";
import subscribe from "../steps";
import { errorLogger } from "../utils";
import { emitter } from "../utils/eventEmitter";
import { Options } from "../steps/types";

export default async ({ name, typescript, useCwd }) => {
  const options: Options = {};

  if (!name) {
    const questions: Question[] = [
      {
        name: "name",
        type: "input",
        message: "Enter a name for the project:",
        default: "my-frontity-project"
      },
      {
        name: "theme",
        type: "input",
        message: "Enter a starter theme to clone:",
        default: "@frontity/mars-theme"
      }
    ];

    const answers = await prompt(questions);
    options.name = answers.name;
    options.theme = answers.theme;
    console.log();
  } else {
    options.name = name;
  }

  options.typescript = typescript;
  options.path = useCwd ? process.cwd() : resolve(process.cwd(), options.name);

  emitter.on("cli:create:error", errorLogger);
  emitter.on("cli:create:message", (message, action) => {
    if (action) ora.promise(action, message);
    else console.log(message);
  });

  await create(options);

  console.log(chalk.bold("\nFrontity project created.\n"));

  const subscribeQuestions: Question[] = [
    {
      name: "subscribe",
      type: "confirm",
      message: "Do you want to receive framework updates by email?",
      default: false
    },
    {
      name: "email",
      type: "input",
      message: "Please, enter your email:",
      when: answers => answers.subscribe
    }
  ];
  const answers = await prompt(subscribeQuestions);

  if (answers.subscribe) {
    console.log();

    emitter.on("cli:create:subscribe", (message, action) => {
      if (action) ora.promise(action, message);
      else console.log(message);
    });

    await subscribe(answers.email);

    console.log("\nThanks for subscribing! ðŸ˜ƒ");
  } else {
    console.log(
      `\nOk, that's fine! ðŸ˜‰\nYou can subscribe at any point with ${chalk.bold.green(
        "npx frontity subscribe <email>"
      )}.`
    );
  }

  console.log(
    `\nRun ${chalk.bold.green(
      `cd ${options.name} && npx frontity dev`
    )} and have fun! ðŸŽ‰\n\nYou can find docs at ${chalk.underline.magenta(
      "https://docs.frontity.org/"
    )}.\nIf you have any doubts, join our community at ${chalk.underline.magenta(
      "https://community.frontity.org/"
    )}.\n`
  );
};
