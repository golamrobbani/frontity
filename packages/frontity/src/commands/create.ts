import chalk from "chalk";
import { emitter } from "../utils/eventEmitter";
import { Options } from "../steps/create/types";
import {
  normalizeOptions,
  ensureProjectDir,
  createPackageJson,
  createFrontitySettings,
  cloneStarterTheme,
  installDependencies,
  downloadFavicon,
  revertProgress
} from "../steps/create";

const defaultOptions: Options = {
  path: process.cwd(),
  typescript: false,
  packages: [
    // "@frontity/wp-source"
  ],
  theme: "@frontity/mars-theme"
};

const emit = (message: string, step?: Promise<void>) => {
  emitter.emit("cli:create:message", message, step);
};

export default async (passedOptions?: Options) => {
  let options: Options;
  let step: Promise<any>;
  let dirExisted: boolean;

  process.on("SIGINT", async () => {
    if (typeof dirExisted !== "undefined")
      await revertProgress(dirExisted, options);
  });

  try {
    // 1. Parses and validates options.
    options = normalizeOptions(defaultOptions, passedOptions);

    const { name, theme, path } = options;

    // 2. Ensures that the project dir exists and is empty.
    step = ensureProjectDir(path);
    emit(`Ensuring ${chalk.yellow(options.path)} directory.`, step);
    dirExisted = await step;

    // 3. Creates `package.json`.
    step = createPackageJson(name, theme, path);
    emit(`Creating ${chalk.yellow("package.json")}.`, step);
    await step;

    // 4. Creates `frontity.settings`.
    const extension = options.typescript ? "ts" : "js";
    step = createFrontitySettings(extension, name, path);
    emit(`Creating ${chalk.yellow(`frontity.settings.${extension}`)}.`, step);
    await step;

    // 5. Clones `@frontity/mars-theme` inside `packages`.
    step = cloneStarterTheme(options);
    emit(`Cloning ${chalk.green(options.theme)}.`, step);
    await step;

    // 6. Installs dependencies.
    step = installDependencies(options);
    emit(`Installing dependencies.`, step);
    await step;

    // 7. Download favicon.
    step = downloadFavicon(options);
    emit(`Downloading ${chalk.yellow("favicon.ico")}.`, step);
    await step;
  } catch (error) {
    if (typeof dirExisted !== "undefined")
      await revertProgress(dirExisted, options);

    if (emitter) emitter.emit("cli:create:error", error);
    else throw error;
  }
};
