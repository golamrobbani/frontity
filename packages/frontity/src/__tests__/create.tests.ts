import create from "../commands/create";
import * as steps from "../steps";
import { emitter } from "../utils/eventEmitter";

jest.mock("../steps");
jest.mock("../utils/eventEmitter");

const mockedSteps = steps as jest.Mocked<typeof steps>;

describe("create", () => {
  beforeEach(() => {
    mockedSteps.normalizeOptions.mockReset();
    mockedSteps.normalizeOptions.mockImplementation(
      (_defaultOptions, passedOptions) => passedOptions
    );
    mockedSteps.ensureProjectDir.mockReset();
    mockedSteps.createPackageJson.mockReset();
    mockedSteps.createFrontitySettings.mockReset();
    mockedSteps.cloneStarterTheme.mockReset();
    mockedSteps.installDependencies.mockReset();
    mockedSteps.revertProgress.mockReset();
  });

  test("goes through all steps", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };
    await create(options);
    expect(mockedSteps.normalizeOptions.mock.calls[0][1]).toMatchSnapshot();
    expect(mockedSteps.ensureProjectDir.mock.calls).toMatchSnapshot();
    expect(mockedSteps.createPackageJson.mock.calls).toMatchSnapshot();
    expect(mockedSteps.createFrontitySettings.mock.calls).toMatchSnapshot();
    expect(mockedSteps.cloneStarterTheme.mock.calls).toMatchSnapshot();
    expect(mockedSteps.installDependencies.mock.calls).toMatchSnapshot();
  });

  test("works correctly when `options.typescript` is false", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project",
      typescript: false
    };
    await create(options);
    expect(mockedSteps.createFrontitySettings).toHaveBeenCalledWith(
      "js",
      options.name,
      options.path
    );
  });

  test("works correctly when `options.typescript` is true", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project",
      typescript: true
    };
    await create(options);
    expect(mockedSteps.createFrontitySettings).toHaveBeenCalledWith(
      "ts",
      options.name,
      options.path
    );
  });

  test("calls removeProgress on error with dirExisted=true", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };
    mockedSteps.ensureProjectDir.mockResolvedValueOnce(true);

    const error = new Error("Mocked Error");
    mockedSteps.createPackageJson.mockImplementation(() => {
      throw error;
    });

    await create(options);
    expect(emitter.emit).toHaveBeenLastCalledWith("cli:create:error", error);
    expect(mockedSteps.revertProgress).toHaveBeenCalledWith(true, options.path);
  });

  test("calls removeProgress on error with dirExisted=false", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };
    mockedSteps.ensureProjectDir.mockResolvedValueOnce(false);

    const error = new Error("Mocked Error");
    mockedSteps.createPackageJson.mockImplementation(() => {
      throw error;
    });

    await create(options);
    expect(emitter.emit).toHaveBeenLastCalledWith("cli:create:error", error);
    expect(mockedSteps.revertProgress).toHaveBeenCalledWith(
      false,
      options.path
    );
  });

  test("does not call removeProgress if ensureProjectDir throws error", async () => {
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };

    const error = new Error("Mocked Error");
    mockedSteps.ensureProjectDir.mockImplementation(() => {
      throw error;
    });

    await create(options);
    expect(emitter.emit).toHaveBeenLastCalledWith("cli:create:error", error);
    expect(mockedSteps.revertProgress).not.toHaveBeenCalled();
  });

  test("uses the emitter passed to log messages", async () => {
    const emitter = { emit: jest.fn() };
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };
    await create(options);
    expect(emitter.emit.mock.calls).toMatchSnapshot();
  });

  test("uses the emitter passed to log the errors", async () => {
    const emitter = { emit: jest.fn() };
    const options = {
      name: "random-name",
      path: "/path/to/project"
    };
    mockedSteps.ensureProjectDir.mockImplementation(() => {
      throw new Error("Mocked Error");
    });
    await create(options);
    expect(emitter.emit.mock.calls).toMatchSnapshot();
  });
});
