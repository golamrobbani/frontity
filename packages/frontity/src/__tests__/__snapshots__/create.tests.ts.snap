// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create goes throw all steps 1`] = `
Object {
  "name": "random-name",
  "path": "/path/to/project",
}
`;

exports[`create goes throw all steps 2`] = `
Array [
  Array [
    Object {
      "name": "random-name",
      "path": "/path/to/project",
    },
  ],
]
`;

exports[`create goes throw all steps 3`] = `
Array [
  Array [
    Object {
      "name": "random-name",
      "path": "/path/to/project",
    },
  ],
]
`;

exports[`create goes throw all steps 4`] = `
Array [
  Array [
    "js",
    Object {
      "name": "random-name",
      "path": "/path/to/project",
    },
  ],
]
`;

exports[`create goes throw all steps 5`] = `
Array [
  Array [
    Object {
      "name": "random-name",
      "path": "/path/to/project",
    },
  ],
]
`;

exports[`create goes throw all steps 6`] = `
Array [
  Array [
    Object {
      "name": "random-name",
      "path": "/path/to/project",
    },
  ],
]
`;

exports[`create uses the emitter passed to log messages 1`] = `Array []`;

exports[`create uses the emitter passed to log the errors 1`] = `Array []`;
